src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -106,6 +125,25 @@ public AssetProvider CreateAssetProvider(Checksum solutionChecksum, SolutionAsse
             bool updatePrimaryBranch,
             Func<Solution, ValueTask<T>> implementation,
             CancellationToken cancellationToken)
+        {
+            try
+            {
+                var pinnedSolution = await GetPinnedSolutionAsync(assetProvider, solutionChecksum, workspaceVersion, updatePrimaryBranch, cancellationToken).ConfigureAwait(false);
+                await using (pinnedSolution.ConfigureAwait(false))
+                    return (pinnedSolution.Solution, await implementation(pinnedSolution.Solution).ConfigureAwait(false));
+            }
+            catch (Exception ex) when (FatalError.ReportAndPropagateUnlessCanceled(ex, cancellationToken, ErrorSeverity.Critical))
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+        }
```

Comment:
this is the bridge to from the existing pattern (which works for all non-streaming cases) to the nwe functionality.  It aims to be extremely trivial.  specifically, it gets the pinnedSolution, then immediately puts it in a `using` so it will clean up.  I then runs the client-provided helper to actually do work.